diff -ur shim-16.1-org/data/sbat.csv shim-16.1/data/sbat.csv
--- shim-16.1-org/data/sbat.csv	2025-08-13 13:14:56.000000000 -0700
+++ shim-16.1/data/sbat.csv	2026-03-30 11:19:33.340872100 -0700
@@ -1,2 +1,3 @@
 sbat,1,SBAT Version,sbat,1,https://github.com/rhboot/shim/blob/main/SBAT.md
 shim,4,UEFI shim,shim,1,https://github.com/rhboot/shim
+shim.terabyte,1,TeraByte,UEFI shim,3,https://www.terabyteunlimited.com
diff -ur shim-16.1-org/Make.defaults shim-16.1/Make.defaults
--- shim-16.1-org/Make.defaults	2025-08-13 13:14:56.000000000 -0700
+++ shim-16.1/Make.defaults	2026-03-30 12:48:13.841877800 -0700
@@ -28,7 +28,7 @@
 DEBUGINFO	?= $(prefix)/lib/debug/
 DEBUGSOURCE	?= $(prefix)/src/debug/
 OSLABEL		?= $(EFIDIR)
-DEFAULT_LOADER	?= \\\\grub$(ARCH_SUFFIX).efi
+DEFAULT_LOADER	?= grub$(ARCH_SUFFIX).efi
 DASHJ		?= -j$(shell echo $$(($$(grep -c "^model name" /proc/cpuinfo) + 1)))
 
 ARCH		?= $(shell $(CC) -dumpmachine | cut -f1 -d- | sed s,i[3456789]86,ia32,)
@@ -98,7 +98,8 @@
 endif
 
 DEFINES		= -DDEFAULT_LOADER='L"$(DEFAULT_LOADER)"' \
-		  -DDEFAULT_LOADER_CHAR='"$(DEFAULT_LOADER)"'
+		  -DDEFAULT_LOADER_CHAR='"$(DEFAULT_LOADER)"' \
+		  -DDISABLE_FALLBACK
 
 INCLUDES	= -nostdinc \
 		  -I$(TOPDIR)/Cryptlib -I$(TOPDIR)/Cryptlib/Include \
@@ -153,6 +154,10 @@
 	DEFINES  += -DDISABLE_REMOVABLE_LOAD_OPTIONS
 endif
 
+ifneq ($(origin DISABLE_LOAD_OPTIONS), undefined)
+	DEFINES  += -DDISABLE_LOAD_OPTIONS
+endif
+
 LIB_GCC		= $(shell $(CC) $(ARCH_CFLAGS) -print-libgcc-file-name)
 EFI_LIBS	= -lefi -lgnuefi --start-group Cryptlib/libcryptlib.a Cryptlib/OpenSSL/libopenssl.a --end-group $(LIB_GCC)
 FORMAT		?= --target efi-app-$(ARCH)
diff -ur shim-16.1-org/pe-relocate.c shim-16.1/pe-relocate.c
--- shim-16.1-org/pe-relocate.c	2025-08-13 13:14:56.000000000 -0700
+++ shim-16.1/pe-relocate.c	2026-04-06 15:25:33.321585753 -0700
@@ -315,17 +315,12 @@
 	return 0;
 }
 
-static const UINT16 machine_type =
+// machine_type not used.
 #if defined(__x86_64__)
-	IMAGE_FILE_MACHINE_X64;
 #elif defined(__aarch64__)
-	IMAGE_FILE_MACHINE_ARM64;
 #elif defined(__arm__)
-	IMAGE_FILE_MACHINE_ARMTHUMB_MIXED;
 #elif defined(__i386__) || defined(__i486__) || defined(__i686__)
-	IMAGE_FILE_MACHINE_I386;
 #elif defined(__ia64__)
-	IMAGE_FILE_MACHINE_IA64;
 #else
 #error this architecture is not supported by shim
 #endif
@@ -333,15 +328,8 @@
 static int
 image_is_loadable(EFI_IMAGE_OPTIONAL_HEADER_UNION *PEHdr)
 {
-	/* If the machine type doesn't match the binary, bail, unless
-	 * we're in an allowed 64-on-32 scenario */
-	if (PEHdr->Pe32.FileHeader.Machine != machine_type) {
-		if (!(machine_type == IMAGE_FILE_MACHINE_I386 &&
-		      PEHdr->Pe32.FileHeader.Machine == IMAGE_FILE_MACHINE_X64 &&
-		      allow_64_bit())) {
-			return 0;
-		}
-	}
+	/* removed logic of 64-on-32 because it stops 32-on-64 (ALLOW_32BIT_KERNEL_ON_X64)  *
+	 * and section not needed since allow_64_bit() already has correct logic.            */
 
 	/* If it's not a header type we recognize at all, bail */
 	switch (PEHdr->Pe32Plus.OptionalHeader.Magic) {
diff -ur shim-16.1-org/shim.c shim-16.1/shim.c
--- shim-16.1-org/shim.c	2025-08-13 13:14:56.000000000 -0700
+++ shim-16.1/shim.c	2026-03-30 12:40:14.405533500 -0700
@@ -80,6 +80,12 @@
 }
 
 static int
+#if defined(DISABLE_FALLBACK)
+should_use_fallback(EFI_HANDLE image_handle UNUSED)
+{
+	return 0;
+}
+#else
 should_use_fallback(EFI_HANDLE image_handle)
 {
 	EFI_LOADED_IMAGE *li;
@@ -135,6 +141,7 @@
 
 	return ret;
 }
+#endif
 /*
  * Open the second stage bootloader and read it into a buffer
  */
@@ -508,6 +515,10 @@
 	}
 #endif
 
+#if defined(DISABLE_LOAD_OPTIONS)
+	dprint("Ignoring all boot options");
+	return EFI_SUCCESS;
+#else
 	efi_status = parse_load_options(li);
 	if (EFI_ERROR(efi_status)) {
 		perror (L"Failed to get load options: %r\n", efi_status);
@@ -515,6 +526,7 @@
 	}
 
 	return EFI_SUCCESS;
+#endif
 }
 
 static SHIM_LOCK shim_lock_interface;
